### DISCUSSION TITLE
Displaying path query results in VS Code: alerts doesn't do anything

### DISCUSSION THREAD

**Discussion by adriaanjacobs** [github/securitylab#274] [Q&A] [answered]
I run a path query closely following [the documentation](https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/) for it, yet I can't seem to figure out how to display the path from source to sink in VS Code using the CodeQL extension.

I imported the `DataFlow::PathGraph` module and my query looks like this:
```
from DataFlow::PathNode address, DataFlow::PathNode condjump, AddressToCondJumpConfig config
where config.hasFlowPath(address, condjump)
select address, "taints", condjump
```
Now according to the documentation the select clause should contain 4 columns `element, source, sink, string` but even if I do that it doesn't show the path. 

It is mentioned I should drop down the 'Results' box and hit alerts, but that doesn't do anything. I'm able to click on 'nodes' and 'edges' though, but I can't make sense of what is displayed there and to which query result it is related (a lot of entries there). 

Is there something else I should do / could try? 

I'm using the coreutils mirror on github downloaded from lgtm.com

Full query included below:

```
/**
 * @name address leaking into condjump
 * @kind path-problem
 * @problem.severity warning
 * @id addresstaintscondjump
 */

import cpp
import semmle.code.cpp.dataflow.TaintTracking
import DataFlow::PathGraph
import semmle.code.cpp.controlflow.Dereferenced

predicate isAddressType(Type type) {
    type instanceof PointerType or type instanceof ArrayType
}

predicate isNullCheck(Expr expr) {
    exists(NotExpr n | expr = n and isAddressType
        (n.getAnOperand().getUnderlyingType()))
}

class AddressToCondJumpConfig extends TaintTracking::Configuration {
    AddressToCondJumpConfig() { this = "AddressToCondJumpConfig" }

    override predicate isSource(DataFlow::Node source) {
        isAddressType
        (source.asExpr().getUnderlyingType())
    }

    override predicate isSanitizer(DataFlow::Node intermediate) {
        dereferenced(intermediate.asExpr())
        or exists(SizeofOperator sizeof | sizeof.getAChild+() = intermediate.asExpr()) // I should probably cover more unevaluated contexts, but it's a start
        or intermediate.asExpr() instanceof PointerDiffExpr
        or exists(ArrayExpr arrayaccess | intermediate.asExpr() = arrayaccess
            and not isAddressType(arrayaccess.getUnderlyingType())
        )
        or isNullCheck(intermediate.asExpr())
    }

    override predicate isSink(DataFlow::Node sink) {
        (sink.asExpr().isCondition() 
            or sink.asExpr().getEnclosingStmt() instanceof ConditionalStmt
            or exists(Loop loop | loop.getCondition() = sink.asExpr()))
        and not isAddressType(sink.asExpr().getUnderlyingType())
    }
}

from DataFlow::PathNode address, DataFlow::PathNode condjump, AddressToCondJumpConfig config
where config.hasFlowPath(address, condjump)
select address, "taints", condjump
```

**Comment by intrigus-lgtm** [comment]
Did you add 
```ql
/**
 * ...
 * @kind path-problem
 * ...
 */
```
to the start of your query?
If I remember correctly that is also needed.

**Reply by adriaanjacobs** [reply]
Oh yes, I forgot to mention, that's also there. Could you show a screenshot of what it's supposed to look like? I think it displays all the nodes in the Results view, the ordering just seems off (I can't make sense of it). 

I have included the full query in my edited answer.

**Comment by p0** [comment, marked as answer]
I think your select clause doesn't follow the expectations. You probably want something like this: `select condjump, address, condjump, "tainted by $@", address, address.toString()`. See also [here](https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/#select-clause) for the full details.

We should really display a better error here, rather than silently producing no alerts...

**Reply by adriaanjacobs** [reply]
That still doesn't do the trick unfortunately, it displays the results in a similar manner. Pressing 'alerts' also still doesn't seem to do anything (also doesn't seem to be loading either). 
It also doesn't fill in the $@ parameter. 

Here's what it looks like.

![image](https://user-images.githubusercontent.com/48367267/108258461-52824400-7160-11eb-8f57-965d30974170.png)

**Reply by adriaanjacobs** [reply]
After altering the query like you suggested and running it on lgtm.com though, I do get the desired result: alerts in the code and the 'show path' button (also the $@ gets filled in). It must be something with how I set up the database then I suppose.

I'm also running VS Code on Linux, if that's relevant.

**Reply by adriaanjacobs** [reply]
So what's the definitive rule on how the select clause should look like? The documentation, like you say, mentions 'element, source, sink, string' but apparently that's a minimum? What is the purpose of the last two (address, address.toString()) in the select clause you provided?

**Reply by gagliardetto** [reply]
![2021-02-17_21-35](https://user-images.githubusercontent.com/15271561/108264703-882f2900-7170-11eb-949a-8f321358021f.png)

**Reply by adriaanjacobs** [reply]
Yes, if I click that, nothing happens.

**Reply by gagliardetto** [reply]
What's the output of the "alerts" thing?

**Reply by adriaanjacobs** [reply]
Nothing, I couldn't show you, nothing changes on the screen when I click it, to put it bluntly.

**Reply by gagliardetto** [reply]
Try 

```
select condjump, address, condjump, "tainted by " + address.toString()
```

**Reply by gagliardetto** [reply]
And then go to the "alerts" thing.

**Reply by adriaanjacobs** [reply]
Still nothing, no difference unfortunately.

**Reply by p0** [reply]
This seems to work for me, on a `coreutils` database downloaded from LGTM via the in-VS-code command; for completeness, here is the edited version of your full query:

```ql
/**
 * @name address leaking into condjump
 * @kind path-problem
 * @problem.severity warning
 * @id addresstaintscondjump
 */

import cpp
import semmle.code.cpp.dataflow.TaintTracking
import DataFlow::PathGraph
import semmle.code.cpp.controlflow.Dereferenced

predicate isAddressType(Type type) {
    type instanceof PointerType or type instanceof ArrayType
}

predicate isNullCheck(Expr expr) {
    exists(NotExpr n | expr = n and isAddressType
        (n.getAnOperand().getUnderlyingType()))
}

class AddressToCondJumpConfig extends TaintTracking::Configuration {
    AddressToCondJumpConfig() { this = "AddressToCondJumpConfig" }

    override predicate isSource(DataFlow::Node source) {
        isAddressType
        (source.asExpr().getUnderlyingType())
    }

    override predicate isSanitizer(DataFlow::Node intermediate) {
        dereferenced(intermediate.asExpr())
        or exists(SizeofOperator sizeof | sizeof.getAChild+() = intermediate.asExpr()) // I should probably cover more unevaluated contexts, but it's a start
        or intermediate.asExpr() instanceof PointerDiffExpr
        or exists(ArrayExpr arrayaccess | intermediate.asExpr() = arrayaccess
            and not isAddressType(arrayaccess.getUnderlyingType())
        )
        or isNullCheck(intermediate.asExpr())
    }

    override predicate isSink(DataFlow::Node sink) {
        (sink.asExpr().isCondition() 
            or sink.asExpr().getEnclosingStmt() instanceof ConditionalStmt
            or exists(Loop loop | loop.getCondition() = sink.asExpr()))
        and not isAddressType(sink.asExpr().getUnderlyingType())
    }
}

from DataFlow::PathNode address, DataFlow::PathNode condjump, AddressToCondJumpConfig config
where config.hasFlowPath(address, condjump)
select condjump, address, condjump, "tainted by $@", address, address.toString()
```
![image](https://user-images.githubusercontent.com/54942558/108265828-3632d700-7161-11eb-9e3e-be8bebb135b3.png)

**Reply by adriaanjacobs** [reply]
I've never seen that screen before. I'll try downloading the database through the VS Code extension now as well (I manually downloaded from LGTM before)

**Reply by gagliardetto** [reply]
How about 

```
select condjump.getNode(), address, condjump, "tainted by something"
```

and then you rerun the query.

**Reply by p0** [reply]
> So what's the definitive rule on how the select clause should look like? The documentation, like you say, mentions 'element, source, sink, string' but apparently that's a minimum? What is the purpose of the last two (address, address.toString()) in the select clause you provided?

The expected format of the `select` is determined by the `@kind` of the query. For `@kind problem`, the `select` has to name a program element and a string that should be displayed as an alert at that program element's location. For `@kind path-problem`, the format is similar, but there is an additional pair of program elements after the first one which indicate the start and end points of a candidate path -- thus, the format of `select $primaryLocation, $pathStart, $pathEnd, $message` is valid.

In both cases, however, you can interpolate placeholders into the message by including `$@` into the strings. For each such placeholder, you should include an additional (program-element, link-display-text) pair in the selection. My screenshot shows what that would look like -- it allows you to click both onto the start (via the in-message link) and end (via the location at the right of the table view) of the path, without expanding the path.

**Reply by p0** [reply]
@adriaanjacobs Could you check if there's any relevant output if you run the "Output: Focus on Output View" command and select "CodeQL Extension Log" in the drop-down of the pane that appears? That might show any relevant errors.

**Reply by intrigus-lgtm** [reply]
Looking at your first screenshot @adriaanjacobs it looks like you are using a pretty old CodeQL Cli version.
Did you try updating the both the CodeQL Cli version and the VS Code codeql extension?

**Reply by adriaanjacobs** [reply]
I get a few errors upon downloading, here's one when I try to open a quick query:

![image](https://user-images.githubusercontent.com/48367267/108266633-9d08be00-716a-11eb-8bf7-9134974775c2.png)

**Reply by adriaanjacobs** [reply]
I'll try updating now, thanks

**Reply by p0** [reply]
Those errors are definitely due to your extension being too old to understand the latest LGTM metadata, but I would have expected version 2.3.3 to work fine with the database you built locally -- I'd still be curious about "CodeQL Extension Log" output from that.

But updating is also a good first step; lots of good improvements between 2.3.3 and 2.4.4 :-)

**Reply by adriaanjacobs** [reply]
Updating worked, apparently I was also using an external CodeQL CLI before, now I deleted that and the extension downloaded the latest version itself. I still get an error upon running quick query though:

![image](https://user-images.githubusercontent.com/48367267/108267009-186a6f80-716b-11eb-96ff-775f587ac74a.png)

And here are the logs @p0 
```
Starting GitHub.vscode-codeql extension
INFO: Could not find CodeQL on path.
INFO: Could not find CodeQL on path.
Searching for latest release including codeql-linux64.zip.
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Could not find CodeQL on path.
Extracting CodeQL CLI to /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1
CodeQL CLI updated to version "v2.4.4".
Checking CodeQL version using CodeQL CLI: /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/codeql version -v --log-to-stderr --format=terse...
[2021-02-17 21:55:15] This is codeql version -v --log-to-stderr --format=terse
[2021-02-17 21:55:15] Terminating normally.

CLI command succeeded.
Found compatible version of CodeQL CLI (version 2.4.4)
Initializing configuration listener...
Initializing CodeQL cli server...
Initializing query server client.
Checking CodeQL version using CodeQL CLI: /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/codeql version -v --log-to-stderr --format=terse...
Starting CodeQL CLI Server using CodeQL CLI: /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/codeql execute cli-server -v --log-to-stderr
CodeQL CLI Server started on PID: 502505
Resolving RAM settings using CodeQL CLI: resolve ram -v --log-to-stderr --format json...
[2021-02-17 21:55:16] This is codeql version -v --log-to-stderr --format=terse
[2021-02-17 21:55:16] Terminating normally.

CLI command succeeded.
CLI command succeeded.
[2021-02-17 21:55:16] This is codeql execute cli-server -v --log-to-stderr

Initializing database manager.
Initializing database panel.
Registering database panel commands.
Initializing query history manager.
Registering query history panel commands.
Initializing results panel interface.
Initializing compare panel interface.
Initializing source archive filesystem provider.
Initializing CodeQL language server.
Initializing QLTest interface.
Registering CodeQL test panel commands.
Registering top-level command palette commands.
Starting language server.
Registering jump-to-definition handlers.
Removing orphaned databases from workspace storage.
Successfully finished extension initialization.
Resolving database using CodeQL CLI: resolve database -v --log-to-stderr --format json /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/cpp/coreutils_coreutils_bd6c97d...
No orphaned databases found.
CLI command succeeded.

Resolving tests using CodeQL CLI: resolve tests --strict-test-discovery -v --log-to-stderr --format json /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries...
CLI command succeeded.

Searching for latest release including codeql-linux64.zip.
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
INFO: Ignoring a release with no assets named codeql-linux64.zip
CodeQL CLI already up to date.
Resolving database using CodeQL CLI: resolve database -v --log-to-stderr --format json /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/cpp/coreutils_coreutils_bd6c97d...
CLI command succeeded.

Resolving qlpack information using CodeQL CLI: resolve qlpacks -v --log-to-stderr --format json --additional-packs /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries...
CLI command succeeded.
[2021-02-17 21:55:34] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries/
[2021-02-17 21:55:34] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/.codeqlmanifest.json
[2021-02-17 21:55:34] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/
[2021-02-17 21:55:34] [DETAILS] resolve qlpacks> Not scanning from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql which is already done.

Could not find qlpack file for dbscheme /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/cpp/coreutils_coreutils_bd6c97d/db-cpp/semmlecode.cpp.dbscheme
Resolving qlpack information using CodeQL CLI: resolve qlpacks -v --log-to-stderr --format json --additional-packs /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries...
CLI command succeeded.
[2021-02-17 21:55:50] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries/
[2021-02-17 21:55:50] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/.codeqlmanifest.json
[2021-02-17 21:55:50] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/
[2021-02-17 21:55:50] [DETAILS] resolve qlpacks> Not scanning from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql which is already done.

Could not find qlpack file for dbscheme /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/cpp/coreutils_coreutils_bd6c97d/db-cpp/semmlecode.cpp.dbscheme
Resolving qlpack information using CodeQL CLI: resolve qlpacks -v --log-to-stderr --format json --additional-packs /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries...
CLI command succeeded.
[2021-02-17 21:56:14] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/workspaceStorage/6b38206bc4bdc194dc5035b485425098/GitHub.vscode-codeql/quick-queries/
[2021-02-17 21:56:14] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql/.codeqlmanifest.json
[2021-02-17 21:56:14] [DETAILS] resolve qlpacks> Scanning for qlpack.yml from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/
[2021-02-17 21:56:14] [DETAILS] resolve qlpacks> Not scanning from /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/distribution1/codeql which is already done.

Could not find qlpack file for dbscheme /home/adriaan/.config/Code/User/globalStorage/github.vscode-codeql/cpp/coreutils_coreutils_bd6c97d/db-cpp/semmlecode.cpp.dbscheme
```

**Reply by p0** [reply]
That's due to the new database from LGTM needing a new checkout of github/codeql. You could either update to the latest `main` for that, or try with your old locally-built database.

**Reply by adriaanjacobs** [reply]
It finally worked! Thank you guys so much! I'm sure I'll have a bunch more questions later (I actually have one more up right now), but this one's fixed, really cool path query feature. Thanks :)

**Reply by adriaanjacobs** [reply]
Unfortunately, when creating the database myself using the CLI (either using 'create' or 'init/trace-command/finalize' the issue persists. I'm using the exact same query as for prepackaged databases, on the exact same codebases, with the exact same CodeQL library folders checkout. Where could I find the logs displaying an error message? Hitting alert doesn't produce anything again.

**Reply by p0** [reply]
That's very odd. Could you share the database somewhere so I can try to reproduce? Any error logs would be in the extension output pane I mentioned earlier.

**Reply by adriaanjacobs** [reply]
I've sent you the project on Slack, I hope that's alright. The extension output log shows this only once, at the very bottom:

'[2021-02-20 13:04:37] CSV_IMB_QUERIES: Query,edges#query#ffffffffffffff nodes#query#fffffffff #select#query#ffffffffffffffffffffff,addresstocondjump.ql,23,Success,30.134,437303,30365
Exception during results interpretation: Reading output of interpretation failed: Error: Cannot create a string longer than 0x1fffffe8 characters. Will show raw results instead.'

**Reply by p0** [reply]
> '[2021-02-20 13:04:37] CSV_IMB_QUERIES: Query,edges#query#ffffffffffffff nodes#query#fffffffff #select#query#ffffffffffffffffffffff,addresstocondjump.ql,23,Success,30.134,437303,30365
> Exception during results interpretation: Reading output of interpretation failed: Error: Cannot create a string longer than 0x1fffffe8 characters. Will show raw results instead.'

Thanks, that's the root cause of the issue -- we're hitting a Node limitation on string length, and the CodeQL extension [is falling back on raw results display](https://github.com/github/vscode-codeql/blob/ab31d86a8d7e09926772e0b84744e8a45639be47/extensions/ql-vscode/src/interface.ts#L630) (which means showing the different query predicates -- `#select`, `edges` and so on -- rather than the alerts view).

As for why we trigger a string length limit: The query as written tuns out to have way too many results on `glibc` (hundreds of thousands). There are some sinks that are tainted by very very many sources, and for such cases the alerts view currently attempts to construct an alert message string that consists of the messages from the `select` for each such source; this is what goes wrong.

To confirm, try running with the following rewrite of the `select` (which arbitrarily restricts to at most 4 sources per sink):

```ql
from
  DataFlow::PathNode address, DataFlow::PathNode condjump, AddressToCondJumpConfig config,
  int numSources, int numToReport, string extra
where
  address =
    rank[[1 .. numToReport]](DataFlow::PathNode node |
      config.hasFlowPath(node, condjump)
    |
      node
      order by
        node.getNode().getLocation().getFile().getAbsolutePath(),
        node.getNode().getLocation().getStartLine()
    ) and
  numSources = count(DataFlow::PathNode node | config.hasFlowPath(node, condjump)) and
  numToReport = 4 and
  if numSources > numToReport
  then extra = " and " + (numSources - numToReport) + " more"
  else extra = ""
select condjump, address, condjump, "tainted by " + address.toString() + extra
```

You should see that this does render in the alerts view.

------

Why are there so many results, and so many source/sink pairs? It turns out the worst offender is in a [macro-generated location](https://github.com/bminor/glibc/blob/83908b3a1ea51e3aa7ff422275940e56dbba989f/iconv/loop.c#L336), which makes it hard to understand the flows. Looking at the query, though, you are asking to track from any expression whose type is in `isAddressType` to any expression that is enclosed in an `if`, or is a condition itself. This is likely to have multiple results for any condition that works with pointers or arrays. I would strongly recommend strengthening your definition of `source`: Perhaps you're interested in function parameters whose types are addresses, or in addresses that come from particular APIs?

------

Still, I think we could do a better job reporting why the `alerts` view is non-functional in your setup. I'll see if we can make that happen.

**Reply by p0** [reply]
I've filed github/vscode-codeql#759 for this.

**Reply by adriaanjacobs** [reply]
Yes, limiting the amount of sources that flow into the same sink has been on my todo-list. 
Thank you so much for helping me get started with that and resolving this issue, as well as providing a detailed explanation of what's going on under the hood.
